// orb-ui/dist/orb-ui.js
import { useCallback as e, useEffect as t, useLayoutEffect as n, useRef as r, useState as i } from "react";
import { jsx as a, jsxs as o } from "react/jsx-runtime";
//#region src/adapters/types.ts
function s(e) {
	let t = Math.min(e / .5, 1);
	return t **= 1.3, t;
}
//#endregion
//#region src/themes/debug/DebugTheme.tsx
var c = [
	"idle",
	"connecting",
	"listening",
	"speaking",
	"error"
], l = {
	idle: "#888",
	connecting: "#f0c040",
	listening: "#40c0f0",
	speaking: "#40f080",
	error: "#f04040"
};
function u({ state: e, volume: t, size: n, className: r, style: i, disabled: s = !1, onStart: u, onStop: d, ...f }) {
	return /* @__PURE__ */ o("div", {
		...f,
		className: r,
		style: {
			width: n,
			fontFamily: "monospace",
			fontSize: 12,
			background: "#111",
			color: "#ccc",
			border: "1px solid #333",
			borderRadius: 8,
			padding: 12,
			boxSizing: "border-box",
			userSelect: "none",
			...i
		},
		children: [
			/* @__PURE__ */ a("div", {
				style: {
					color: "#555",
					marginBottom: 10,
					fontSize: 10,
					letterSpacing: 1
				},
				children: "ORB DEBUG"
			}),
			/* @__PURE__ */ o("div", {
				style: { marginBottom: 8 },
				children: [/* @__PURE__ */ a("span", {
					style: { color: "#555" },
					children: "state "
				}), /* @__PURE__ */ a("span", {
					style: {
						color: l[e],
						fontWeight: "bold"
					},
					children: e
				})]
			}),
			/* @__PURE__ */ o("div", {
				style: { marginBottom: 10 },
				children: [
					/* @__PURE__ */ a("span", {
						style: { color: "#555" },
						children: "volume "
					}),
					/* @__PURE__ */ a("span", {
						style: { color: "#ccc" },
						children: t.toFixed(2)
					}),
					/* @__PURE__ */ a("div", {
						style: {
							marginTop: 4,
							height: 4,
							background: "#222",
							borderRadius: 2,
							overflow: "hidden"
						},
						children: /* @__PURE__ */ a("div", { style: {
							height: "100%",
							width: `${t * 100}%`,
							background: l[e],
							borderRadius: 2,
							transition: "width 50ms linear"
						} })
					})
				]
			}),
			/* @__PURE__ */ o("div", {
				style: { marginBottom: 10 },
				children: [/* @__PURE__ */ a("div", {
					style: {
						color: "#555",
						marginBottom: 4,
						fontSize: 10
					},
					children: "force state"
				}), /* @__PURE__ */ a("div", {
					style: {
						display: "flex",
						flexWrap: "wrap",
						gap: 4
					},
					children: c.map((t) => /* @__PURE__ */ a("button", {
						disabled: s,
						style: {
							fontSize: 10,
							padding: "2px 6px",
							background: e === t ? l[t] : "#222",
							color: e === t ? "#000" : "#888",
							border: `1px solid ${e === t ? l[t] : "#333"}`,
							borderRadius: 3,
							cursor: s ? "not-allowed" : "pointer"
						},
						onClick: () => {
							console.warn(`[orb-ui debug] To force state '${t}', use controlled mode: <Orb state="${t}" />`);
						},
						children: t
					}, t))
				})]
			}),
			/* @__PURE__ */ o("div", {
				style: {
					display: "flex",
					gap: 6
				},
				children: [/* @__PURE__ */ a("button", {
					disabled: s,
					onClick: u,
					style: {
						flex: 1,
						padding: "4px 0",
						background: "#1a3a1a",
						color: "#40f080",
						border: "1px solid #40f080",
						borderRadius: 4,
						cursor: s ? "not-allowed" : "pointer",
						fontSize: 11
					},
					children: "Start"
				}), /* @__PURE__ */ a("button", {
					disabled: s,
					onClick: d,
					style: {
						flex: 1,
						padding: "4px 0",
						background: "#3a1a1a",
						color: "#f04040",
						border: "1px solid #f04040",
						borderRadius: 4,
						cursor: s ? "not-allowed" : "pointer",
						fontSize: 11
					},
					children: "Stop"
				})]
			})
		]
	});
}
//#endregion
//#region src/themes/circle/CircleTheme.tsx
var d = typeof window < "u" ? n : t;
function f(e) {
	return [
		parseInt(e.slice(1, 3), 16),
		parseInt(e.slice(3, 5), 16),
		parseInt(e.slice(5, 7), 16)
	];
}
var p = {
	idle: "#cccccc",
	connecting: "#cccccc",
	listening: "#999999",
	speaking: "#e8e8e8",
	error: "#f87171"
}, m = "\n@keyframes orb-circle-idle-pulse {\n  from { transform: scale(1); }\n  to   { transform: scale(1.06); }\n}\n@keyframes orb-circle-connecting-pulse {\n  0%   { opacity: 1; transform: scale(1); }\n  50%  { opacity: 0.6; transform: scale(0.95); }\n  100% { opacity: 1; transform: scale(1); }\n}\n", h = .95, g = .08, _ = .82, v = .18, y = 24, b = 0, x = .55, S = .12, C = .002;
function w({ state: e, volume: n, size: i, className: s, style: c, disabled: l = !1, interactive: u = !1, onClick: w, ...T }) {
	let E = r(null), D = r(null), O = r(null), k = r(0), A = r(n);
	d(() => {
		A.current = n;
	}, [n]);
	let j = r(1), M = r(0), N = r(f(p.idle)), P = .06, F = r(_), I = r(v);
	t(() => {
		let e = "orb-circle-keyframes";
		if (!document.getElementById(e)) {
			let t = document.createElement("style");
			t.id = e, t.textContent = m, document.head.appendChild(t);
		}
	}, []), t(() => {
		let t = E.current;
		if (t) if (e === "listening" || e === "speaking") {
			let n = e === "speaking" ? h : _, r = e === "speaking" ? g : v, i = e === "speaking" ? y : b, a = () => {
				let o = A.current;
				F.current += (n - F.current) * P, I.current += (r - I.current) * P;
				let s = F.current + o * I.current, c = o * i;
				e === "listening" ? (j.current += (s - j.current) * x, M.current += (c - M.current) * x) : (j.current = s, M.current = c);
				let l = f(p[e]), [u, d, m] = N.current;
				N.current = [
					u + (l[0] - u) * .05,
					d + (l[1] - d) * .05,
					m + (l[2] - m) * .05
				];
				let [h, g, _] = N.current.map(Math.round);
				t.style.transform = `scale(${j.current})`, t.style.background = `rgb(${h},${g},${_})`, t.style.boxShadow = "none", t.style.animation = "none";
				let v = D.current;
				if (v) {
					let e = M.current;
					v.style.transform = `scale(${j.current})`, v.style.boxShadow = e > .5 ? `0 0 ${e}px ${e * .4}px rgb(${h},${g},${_})` : "none";
				}
				k.current = requestAnimationFrame(a);
			};
			return k.current = requestAnimationFrame(a), () => {
				cancelAnimationFrame(k.current);
			};
		} else {
			cancelAnimationFrame(k.current);
			let n = p[e] ?? p.idle, r = f(n), i = () => {
				j.current += (1 - j.current) * S, M.current += (0 - M.current) * S;
				let [a, o, s] = N.current;
				N.current = [
					a + (r[0] - a) * S,
					o + (r[1] - o) * S,
					s + (r[2] - s) * S
				];
				let [c, l, u] = N.current.map(Math.round);
				t.style.transform = `scale(${j.current})`, t.style.background = `rgb(${c},${l},${u})`, t.style.boxShadow = "none", t.style.animation = "none", D.current && (D.current.style.transform = `scale(${j.current})`, D.current.style.boxShadow = "none");
				let d = Math.abs(j.current - 1) < C, f = M.current < .1, p = N.current.every((e, t) => Math.abs(e - r[t]) < 1);
				if (d && f && p) {
					j.current = 1, M.current = 0, N.current = r, t.style.transform = "", t.style.boxShadow = "none", t.style.background = n, D.current && (D.current.style.transform = "scale(1)", D.current.style.boxShadow = "none"), e === "idle" ? t.style.animation = "orb-circle-idle-pulse 3s ease-in-out infinite alternate" : e === "connecting" ? t.style.animation = "orb-circle-connecting-pulse 1.5s ease-in-out infinite" : t.style.animation = "none";
					return;
				}
				k.current = requestAnimationFrame(i);
			};
			return k.current = requestAnimationFrame(i), () => cancelAnimationFrame(k.current);
		}
	}, [e]);
	let L = i * .55, R = {
		width: i,
		height: i,
		display: "flex",
		alignItems: "center",
		justifyContent: "center",
		position: "relative",
		...c
	}, z = /* @__PURE__ */ o("span", {
		ref: O,
		onMouseEnter: () => {
			O.current && !l && (O.current.style.transform = "scale(1.06)", O.current.style.filter = "brightness(1.12)");
		},
		onMouseLeave: () => {
			O.current && (O.current.style.transform = "scale(1)", O.current.style.filter = "brightness(1)");
		},
		onTouchEnd: () => {
			setTimeout(() => {
				O.current && (O.current.style.transform = "scale(1)", O.current.style.filter = "brightness(1)");
			}, 200);
		},
		style: {
			position: "relative",
			display: "inline-block",
			transition: "transform 0.3s ease, filter 0.3s ease",
			cursor: u ? l ? "not-allowed" : "pointer" : "default",
			borderRadius: "50%",
			lineHeight: 0
		},
		children: [/* @__PURE__ */ a("span", {
			ref: D,
			style: {
				position: "absolute",
				display: "block",
				width: L,
				height: L,
				borderRadius: "50%",
				pointerEvents: "none"
			}
		}), /* @__PURE__ */ a("span", {
			ref: E,
			style: {
				position: "relative",
				display: "block",
				width: L,
				height: L,
				borderRadius: "50%",
				background: p[e]
			}
		})]
	});
	return u ? /* @__PURE__ */ a("button", {
		...T,
		type: "button",
		className: s,
		disabled: l,
		onClick: l ? void 0 : w,
		style: {
			appearance: "none",
			WebkitAppearance: "none",
			border: 0,
			padding: 0,
			margin: 0,
			background: "transparent",
			color: "inherit",
			font: "inherit",
			cursor: l ? "not-allowed" : "pointer",
			...R
		},
		children: z
	}) : /* @__PURE__ */ a("div", {
		...T,
		className: s,
		style: R,
		children: z
	});
}
//#endregion
//#region src/themes/bars/BarsTheme.tsx
var T = 5, E = 1.4, D = Math.PI * 2 / T, O = {
	idle: "#cccccc",
	connecting: "#cccccc",
	listening: "#999999",
	speaking: "#e8e8e8",
	error: "#f87171"
};
function k(e) {
	return [
		parseInt(e.slice(1, 3), 16),
		parseInt(e.slice(3, 5), 16),
		parseInt(e.slice(5, 7), 16)
	];
}
function A({ state: e, volume: n, size: i, className: o, style: s, disabled: c = !1, interactive: l = !1, onClick: u, ...d }) {
	let f = r([]), p = r(null), m = r(0), h = r(Array(T).fill(0)), g = r(n), _ = r(!1), v = r(0), y = r(k(O.idle)), b = r(null), x = r(Array(T).fill(0)), S = r(e);
	t(() => {
		e !== S.current && (x.current = [...h.current], b.current = Date.now(), S.current = e);
	}, [e]), t(() => {
		g.current = n;
	}, [n]), t(() => {
		let t = i * .55, n = i * .06, r = O[e] ?? O.idle, a = i * .1, o = Array.from({ length: T }, (e, t) => {
			let n = (T - 1) / 2;
			return 1 - .7 * (Math.abs(t - n) / n);
		}), s = () => {
			let e = _.current ? a : 0;
			v.current += (e - v.current) * .15;
		}, c = (n, r) => {
			s();
			let i = k(r), [a, c, l] = y.current;
			y.current = [
				a + (i[0] - a) * .08,
				c + (i[1] - c) * .08,
				l + (i[2] - l) * .08
			];
			let [u, d, p] = y.current.map(Math.round), m = `rgb(${u},${d},${p})`;
			for (let r = 0; r < T; r++) {
				let i = f.current[r];
				if (!i) continue;
				let a = e === "idle" ? o[r] : 1, s = v.current * a;
				i.style.height = `${Math.min(n[r] + s, t)}px`, i.style.background = m, i.style.animation = "none";
			}
		};
		if (e === "listening" || e === "speaking") {
			let i = e === "speaking" ? 1 : .4, a = () => {
				let o = g.current, s = Date.now() / 1e3;
				for (let r = 0; r < T; r++) {
					let a = .5 + .15 * Math.sin(s * E * i * Math.PI * 2 + r * D), c = n + (t - n) * o * a;
					if (b.current !== null) {
						let e = Date.now() - b.current, t = Math.min(e / 300, 1), n = 1 - (1 - t) * (1 - t);
						c = x.current[r] + (c - x.current[r]) * n, t >= 1 && (b.current = null);
					}
					let l = e === "listening" ? .45 : 1;
					h.current[r] += (c - h.current[r]) * l;
				}
				c(h.current, r), m.current = requestAnimationFrame(a);
			};
			return m.current = requestAnimationFrame(a), () => cancelAnimationFrame(m.current);
		}
		if (e === "connecting") {
			let e = Date.now(), i = () => {
				let a = (Date.now() - e) / 1e3;
				s();
				for (let e = 0; e < T; e++) {
					let r = (a * .6 + e / T * .5) % 1, i = r < .5 ? Math.sin(r / .5 * Math.PI) : 0, o = n + (t * .4 - n) * i;
					h.current[e] += (o - h.current[e]) * .15;
				}
				c(h.current, r), m.current = requestAnimationFrame(i);
			};
			return m.current = requestAnimationFrame(i), () => cancelAnimationFrame(m.current);
		}
		cancelAnimationFrame(m.current);
		let l = () => {
			s();
			for (let e = 0; e < T; e++) h.current[e] += (n - h.current[e]) * .16;
			c(h.current, r), m.current = requestAnimationFrame(l);
		};
		return m.current = requestAnimationFrame(l), () => cancelAnimationFrame(m.current);
	}, [e, i]);
	let C = i * .055, w = i * .035, A = i * .03, j = i * .55, M = i * .06, N = {
		width: i,
		height: i,
		display: "flex",
		alignItems: "center",
		justifyContent: "center",
		position: "relative",
		...s
	}, P = /* @__PURE__ */ a("span", {
		ref: p,
		onMouseEnter: () => {
			c || (_.current = !0, p.current && (p.current.style.filter = "brightness(1.35)"));
		},
		onMouseLeave: () => {
			_.current = !1, p.current && (p.current.style.filter = "brightness(1)");
		},
		onTouchEnd: () => {
			setTimeout(() => {
				_.current = !1, p.current && (p.current.style.filter = "brightness(1)");
			}, 200);
		},
		style: {
			display: "flex",
			alignItems: "center",
			justifyContent: "center",
			gap: w,
			transition: "filter 0.3s ease",
			cursor: l ? c ? "not-allowed" : "pointer" : "default"
		},
		children: Array.from({ length: T }, (t, n) => /* @__PURE__ */ a("span", {
			ref: (e) => {
				f.current[n] = e;
			},
			style: {
				width: C,
				minHeight: M,
				maxHeight: j,
				height: M,
				borderRadius: A,
				background: O[e] ?? O.idle
			}
		}, n))
	});
	return l ? /* @__PURE__ */ a("button", {
		...d,
		type: "button",
		className: o,
		disabled: c,
		onClick: c ? void 0 : u,
		style: {
			appearance: "none",
			WebkitAppearance: "none",
			border: 0,
			padding: 0,
			margin: 0,
			background: "transparent",
			color: "inherit",
			font: "inherit",
			cursor: c ? "not-allowed" : "pointer",
			...N
		},
		children: P
	}) : /* @__PURE__ */ a("div", {
		...d,
		className: o,
		style: N,
		children: P
	});
}
//#endregion
//#region src/components/Orb/Orb.tsx
function j() {
	let e = null, t = null, n = null, r = 0, i = 0, a = null;
	if (typeof navigator < "u" && navigator.mediaDevices?.getUserMedia) {
		let e = navigator.mediaDevices.getUserMedia.bind(navigator.mediaDevices);
		navigator.mediaDevices.getUserMedia = async (t) => {
			let n = await e(t);
			return t?.audio && (a = n), n;
		};
	}
	function o(o) {
		if (!(!a || t)) try {
			e = new AudioContext(), t = e.createAnalyser(), t.fftSize = 256, n = e.createMediaStreamSource(a), n.connect(t);
			let c = new Uint8Array(t.frequencyBinCount), l = () => {
				if (!t) return;
				t.getByteFrequencyData(c);
				let e = 0;
				for (let t = 0; t < c.length; t++) e += c[t] * c[t];
				let n = Math.sqrt(e / c.length) / 255, a = n > i ? .7 : .3;
				i += (n - i) * a, o(s(i)), r = requestAnimationFrame(l);
			};
			r = requestAnimationFrame(l);
		} catch (e) {
			console.warn("[orb-ui] Mic monitor failed:", e);
		}
	}
	function c() {
		cancelAnimationFrame(r), r = 0, i = 0, n?.disconnect(), n = null, t = null;
	}
	function l() {
		c(), e &&= (e.close().catch(() => {}), null), a?.getTracks().forEach((e) => {
			e.readyState === "live" && e.stop();
		}), a = null;
	}
	return {
		start: o,
		stop: c,
		release: l
	};
}
function M({ state: n, volume: o, adapter: s, theme: c = "debug", size: l = 200, className: d, style: f, disabled: p = !1, onStart: m, onStop: h, ...g }) {
	let [_, v] = i("idle"), [y, b] = i(0), [x, S] = i(0), C = r(j()), T = r(!1);
	t(() => {
		if (v("idle"), b(0), S(0), s) return s.subscribe({
			onStateChange: v,
			onVolumeChange: b
		});
	}, [s]);
	let E = n ?? _;
	t(() => {
		let e = C.current;
		if (E === "listening" && s) {
			let t = setTimeout(() => {
				e.start(S), T.current = !0;
			}, 500);
			return () => {
				clearTimeout(t), e.stop(), T.current = !1, S(0);
			};
		} else e.stop(), T.current = !1, S(0);
	}, [E, s]), t(() => () => C.current.release(), []);
	let D = o ?? (E === "listening" && T.current ? x : y), O = E !== "idle" && E !== "error", k = e(() => {
		p || (O ? h ? h() : s?.stop?.() : m ? m() : s?.start?.());
	}, [
		s,
		p,
		O,
		m,
		h
	]), M = O ? !!(s?.stop || h) : !!(s?.start || m), N = M && !p ? k : void 0, P = {
		state: E,
		volume: D,
		size: l,
		className: d,
		style: f,
		disabled: p,
		...g,
		"aria-label": g["aria-label"] ?? (M ? `${O ? "Stop" : "Start"} voice session` : void 0)
	}, F = {
		...P,
		interactive: M
	};
	switch (c) {
		case "circle": return /* @__PURE__ */ a(w, {
			...F,
			onClick: N
		});
		case "bars": return /* @__PURE__ */ a(A, {
			...F,
			onClick: N
		});
		default: return /* @__PURE__ */ a(u, {
			...P,
			onStart: p ? void 0 : m ?? (() => s?.start?.()),
			onStop: p ? void 0 : h ?? (() => s?.stop?.())
		});
	}
}
//#endregion
export { M as Orb, M as VoiceOrb };
